ttljob: wire up range_concurrency and cleanup query construction

otan opened the pull request:
See individual commits for details.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76487)
<!-- Reviewable:end -->

otan commented on pkg/sql/ttl/ttljob/ttljob_test.go:
```diff
@@ -8,33 +8,348 @@
+			numNonExpiredRows: 5,
+		},
+		{
+			desc: "one column pk, concurrent",
```
this file was a move from `ttljob_integration_test.go`; only the concurrent tests are new

otan committed:
sql: add syntax to configure ttl_range_concurrency

This commit adds the ability to configure how many ranges to scan at
once for each TTL job.

Release note: None

otan committed:
ttljob: wire up range_concurrency and cleanup query construction

This commit changes the TTL job to run one range at a time, with the
option of running with concurrency.

We also split the query generation into builders, which caches query
strings which may be continually re-used as well as re-using the same
buffer space for arg allocations. This saves some CPU time.

Release note: None

rafiss reviewed (changes requested):
some minor nits

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76487#-:-MvzAgKI6YROUWeHvTsc:b2ohgjw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[pkg/sql/paramparse/paramobserver.go, line 323 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76487#-Mvz5brLEV-ohoJrJAOn:-Mvz5brLEV-ohoJrJAOo:bvqarpk) ([raw file](https://github.com/cockroachdb/cockroach/blob/22eb8deeaec27ba06490f0e4f09e11582be0a8c2/pkg/sql/paramparse/paramobserver.go#L323)):*
> ```Go
> 		onReset: func(po *TableStorageParamObserver, evalCtx *tree.EvalContext, key string) error {
> 			if po.tableDesc.RowLevelTTL != nil {
> 				po.tableDesc.RowLevelTTL.RangeConcurrency = 0
> ```

should it reset to 1? (and if so, needs a test)
___
*[pkg/sql/ttl/ttljob/ttljob.go, line 110 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76487#-Mvz5ts7By9pLOd-vHqv:-Mvz5ts7By9pLOd-vHqw:b-ltxify) ([raw file](https://github.com/cockroachdb/cockroach/blob/22eb8deeaec27ba06490f0e4f09e11582be0a8c2/pkg/sql/ttl/ttljob/ttljob.go#L110)):*
> ```Go
> 		// early as the delete will not work.
> 		if desc.GetModificationTime().GoTime().After(aost.Time) {
> 			return errors.Newf("found a recent schema change on the table, aborting")
> ```

is there any use in including the timestamps in the error message?
___
*[pkg/sql/ttl/ttljob/ttljob.go, line 142 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76487#-Mvz8dwp27o6ybfEHBOD:-Mvz8dwp27o6ybfEHBOE:b3cnbne) ([raw file](https://github.com/cockroachdb/cockroach/blob/22eb8deeaec27ba06490f0e4f09e11582be0a8c2/pkg/sql/ttl/ttljob/ttljob.go#L142)):*
> ```Go
> 		startPK, endPK tree.Datums
> 	}
> 	ch := make(chan rangeToProcess)
> ```

should this channel be size-capped at the level of concurrency?
___
*[pkg/sql/ttl/ttljob/ttljob.go, line 178 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76487#-Mvz8Ah_9TDW6Edh0lIe:-Mvz8Ah_9TDW6Edh0lIf:b49rlek) ([raw file](https://github.com/cockroachdb/cockroach/blob/22eb8deeaec27ba06490f0e4f09e11582be0a8c2/pkg/sql/ttl/ttljob/ttljob.go#L178)):*
> ```Go
> 
> 	if err := func() error {
> 		defer close(ch)
> ```

i feel like it might be nicer to not have to call `g.Wait` in two separate places. what about

```
func() (retErr error) {
  defer func() { retErr = errors.CombineErrors(retErr, g.Wait()) }()
  defer close(ch)
....
```
___
*[pkg/sql/ttl/ttljob/ttljob.go, line 183 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76487#-Mvz7i4w4HgJSNtRSIme:-Mvz7i4w4HgJSNtRSImf:b-5gn9vj) ([raw file](https://github.com/cockroachdb/cockroach/blob/22eb8deeaec27ba06490f0e4f09e11582be0a8c2/pkg/sql/ttl/ttljob/ttljob.go#L183)):*
> ```Go
> 				return err
> 			}
> 			var r rangeToProcess
> ```

is `r` shadowed here on purpose? if not, would be good to use some descriptive names
___
*[pkg/sql/ttl/ttljob/ttljob_query_builder.go, line 36 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76487#-Mvz9csy5Ad1yW_5kryZ:-Mvz9csy5Ad1yW_5kry_:b-x9qb5l) ([raw file](https://github.com/cockroachdb/cockroach/blob/22eb8deeaec27ba06490f0e4f09e11582be0a8c2/pkg/sql/ttl/ttljob/ttljob_query_builder.go#L36)):*
> ```Go
> 	aost            tree.DTimestampTZ
> 
> 	// isFirst is true if we have no invoked a query using the builder yet.
> ```

nit: no->not
___
*[pkg/sql/ttl/ttljob/ttljob_query_builder.go, line 40 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76487#-Mvz9h1g88XxIh2PV1F2:-Mvz9h1g88XxIh2PV1F3:bc62hc5) ([raw file](https://github.com/cockroachdb/cockroach/blob/22eb8deeaec27ba06490f0e4f09e11582be0a8c2/pkg/sql/ttl/ttljob/ttljob_query_builder.go#L40)):*
> ```Go
> 	// queryCache is the cached query, which stays the same from the second
> 	// iteration onwards.
> 	queryCache string
> ```

nit: maybe rename to `cachedQuery` and `cachedArgs`? the phrasing now makes it seem like they are caches, but they actually are cached data
___
*[pkg/sql/ttl/ttljob/ttljob_query_builder_test.go, line 271 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76487#-MvzAJ8j9O_onQivtU_x:-MvzAJ8j9O_onQivtU_y:b-j6rd7h) ([raw file](https://github.com/cockroachdb/cockroach/blob/22eb8deeaec27ba06490f0e4f09e11582be0a8c2/pkg/sql/ttl/ttljob/ttljob_query_builder_test.go#L271)):*
> ```Go
> 	}
> 
> 	for _, tc := range testCases {
> ```

nice tests


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76487)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[pkg/sql/exec_util.go, line 1479 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76487#-MvzAzrDBre0mN9WDj3Y:-MvzAzrDBre0mN9WDj3Z:b-wrqqgp) ([raw file](https://github.com/cockroachdb/cockroach/blob/22eb8deeaec27ba06490f0e4f09e11582be0a8c2/pkg/sql/exec_util.go#L1479)):*
> ```Go
> 	// MockDescriptorVersionDuringDelete is a version to mock the delete descriptor
> 	// as during delete.
> 	MockDescriptorVersionDuringDelete *descpb.DescriptorVersion
> ```

do these two knobs need to be pointers? (or is that a convention?)


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76487)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/sql/exec_util.go, line 1479 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76487#-MvzAzrDBre0mN9WDj3Y:-MvzGhBrA21mMeaea94E:bewv3b8) ([raw file](https://github.com/cockroachdb/cockroach/blob/22eb8deeaec27ba06490f0e4f09e11582be0a8c2/pkg/sql/exec_util.go#L1479)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

do these two knobs need to be pointers? (or is that a convention?)
</blockquote></details>

it's to differentiate do not set vs run and set. i've seen it in a few knobs.
___
*[pkg/sql/paramparse/paramobserver.go, line 323 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76487#-Mvz5brLEV-ohoJrJAOn:-MvzFztn4D1QfeEl3Va1:b-gebuh8) ([raw file](https://github.com/cockroachdb/cockroach/blob/22eb8deeaec27ba06490f0e4f09e11582be0a8c2/pkg/sql/paramparse/paramobserver.go#L323)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

should it reset to 1? (and if so, needs a test)
</blockquote></details>

nah it's 0. if it is 0, we use the cluster setting default.
___
*[pkg/sql/ttl/ttljob/ttljob.go, line 110 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76487#-Mvz5ts7By9pLOd-vHqv:-MvzI5XR0GH-bfAqcCzH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/22eb8deeaec27ba06490f0e4f09e11582be0a8c2/pkg/sql/ttl/ttljob/ttljob.go#L110)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

is there any use in including the timestamps in the error message?
</blockquote></details>

Done.
___
*[pkg/sql/ttl/ttljob/ttljob.go, line 142 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76487#-Mvz8dwp27o6ybfEHBOD:-MvzGpNv5rqxKUgC3yQp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/22eb8deeaec27ba06490f0e4f09e11582be0a8c2/pkg/sql/ttl/ttljob/ttljob.go#L142)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

should this channel be size-capped at the level of concurrency?
</blockquote></details>

Done.
___
*[pkg/sql/ttl/ttljob/ttljob.go, line 178 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76487#-Mvz8Ah_9TDW6Edh0lIe:-MvzGwzH26XPl4kSpy98:b-w0jnzl) ([raw file](https://github.com/cockroachdb/cockroach/blob/22eb8deeaec27ba06490f0e4f09e11582be0a8c2/pkg/sql/ttl/ttljob/ttljob.go#L178)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i feel like it might be nicer to not have to call `g.Wait` in two separate places. what about

```
func() (retErr error) {
  defer func() { retErr = errors.CombineErrors(retErr, g.Wait()) }()
  defer close(ch)
....
```
</blockquote></details>

that's more prone to bugs imo. if `retErr` is nil, `CombinedErrors` will be nil even if `g.Wait()` is not nil!
___
*[pkg/sql/ttl/ttljob/ttljob.go, line 183 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76487#-Mvz7i4w4HgJSNtRSIme:-MvzGbSFAMa7AYPy2wOL:b-vfgh6b) ([raw file](https://github.com/cockroachdb/cockroach/blob/22eb8deeaec27ba06490f0e4f09e11582be0a8c2/pkg/sql/ttl/ttljob/ttljob.go#L183)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

is `r` shadowed here on purpose? if not, would be good to use some descriptive names
</blockquote></details>

Done. (Yes but descriptive names are good)
___
*[pkg/sql/ttl/ttljob/ttljob_query_builder_test.go, line 271 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76487#-MvzAJ8j9O_onQivtU_x:-MvzGf-iEcrRpINGz0be:b-rvg99h) ([raw file](https://github.com/cockroachdb/cockroach/blob/22eb8deeaec27ba06490f0e4f09e11582be0a8c2/pkg/sql/ttl/ttljob/ttljob_query_builder_test.go#L271)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nice tests
</blockquote></details>

danke!


<!-- Sent from Reviewable.io -->

otan committed:
ttljob: block job from executing if schema change is detected

Release note: None

rafiss reviewed (approved):
lgtm! still an optional nit on CombineErrors usage

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76487#-:-Mw3q7IN-uG0-qS6GC-R:b7e90xa)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[pkg/sql/ttl/ttljob/ttljob.go, line 178 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76487#-Mvz8Ah_9TDW6Edh0lIe:-Mw3pO8U-g-xFxmRpQtb:b2cgpnm) ([raw file](https://github.com/cockroachdb/cockroach/blob/22eb8deeaec27ba06490f0e4f09e11582be0a8c2/pkg/sql/ttl/ttljob/ttljob.go#L178)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

that's more prone to bugs imo. if `retErr` is nil, `CombinedErrors` will be nil even if `g.Wait()` is not nil!
</blockquote></details>

that's how Wrap works. CombineErrors does what we want
```
// CombineErrors returns err, or, if err is nil, otherErr.
// if err is non-nil, otherErr is attached as secondary error.
```

we rely on this pattern a lot. e.g. https://github.com/cockroachdb/cockroach/blob/49961770c06c1402d1f554d45ee3ea28204ab971/pkg/server/admin.go#L264


<!-- Sent from Reviewable.io -->

otan commented:
done

bors r=rafiss

otan commented:
bors r=rafiss
